### PR TITLE
Introducing Operator types

### PR DESCRIPTION
**Pull request by GeorgeR227**
This PR is driven in part by two different desires. There is a past one, in issue #53, to generalize the inference/overloading rules since there was a lot of repetition. Then there is also the current one, in issue #81, to remove the users from adding in DEC operator annotations. 

My idea is to begin the process of having DEC operators as simply being Symbols in the ACSet to being an Operator node which is richer in that it has typing and naming information attached to it. So for example we'll no longer have `:d` floating around but `:d` that acts on `:Form0` and returns `:Form1` and has other aliases `:d_0, etc...`. This PR won't change the actual ACSet to use this since this will be massively breaking but it will be used heavily for our compilation code.

This data organization moves our ACSet code closer to that of a theory, where operators are not just names but are expected to act on and return certain types. This structure also allows us to present generic names to the user while having functions that can take in the struct information and output a unique name for the operator.

There is some future-proofing included which expects the similar shift from Symbol types, like `:Form0`, to our new Symbolic types. A goal I'd want to tackle but will leave out of scope for when our Symbolics code matures more is to merge these Operators with their Symbolic function representations.

### CODE REVIEW

File: src/acset.jl [hunk 1]
```diff
@@ -554,38 +620,55 @@ function infer_types!(d::SummationDecapode, op1_rules::Vector{NamedTuple{(:src_t
...
-  for op2_idx in parts(d, :Op2)
-    proj1 = d[:proj1][op2_idx]; proj2 = d[:proj2][op2_idx]; res = d[:res][op2_idx]; op2 = d[:op2][op2_idx]
-    proj1_type = d[:type][proj1]; proj2_type = d[:type][proj2]; res_type = d[:type][res]
-    for rule in op2_rules
-      if op2 == rule[:op] && proj1_type == rule[:proj1_type] && proj2_type == rule[:proj2_type] && res_type == rule[:res_type]
-        d[op2_idx, :op2] = rule[:resolved_name]
+  d
+end
+
+function type_check(d::SummationDecapode, type_rules::AbstractVector{Operator{Symbol}})
```
Comment (lukem12345) [inline src/acset.jl:639]: The control flow of this function is hard to parse. If tree probably is called for here.

Comment (lukem12345) [reply src/acset.jl:639]: Actually, the control flow should be reworked such that all type errors are returned.

File: src/acset.jl [hunk 2]
```diff
@@ -440,6 +442,50 @@ function filterfor_ec_types(types::AbstractVector{Symbol})
...
+  map(1:length(types)) do i
+    aliases = isempty(sp_aliases) ? g_aliases : vcat(g_aliases, sp_aliases[i])
+    Operator{Symbol}(types[i], repeat([types[i]], arity), op_name, aliases)
+  end
+end
+
+function arthimetic_operators(op_name::Symbol, broadcasted::Bool, arity::Int = 2)
+  @match (broadcasted, arity) begin
+    (true, 2) => bin_broad_arith_ops(op_name)
+    _ => error("This type of arthimetic operator is not yet supported or may not be valid.")
```
Comment (lukem12345) [inline src/acset.jl:475]: Little typo here in arithmetic

File: src/acset.jl [hunk 3]
```diff
@@ -466,36 +512,99 @@ function infer_sum_types!(d::SummationDecapode, Σ_idx::Int)
...
 
-  check_op = (d[op1_id, :op1] in rule.op_names)
-  if(check_op && (score_src + score_tgt == 1))
-    mod_src = safe_modifytype!(d, d[op1_id, :src], rule.src_type)
-    mod_tgt = safe_modifytype!(d, d[op1_id, :tgt], rule.tgt_type)
-    return mod_src || mod_tgt
+  if name_present && type_diff == 1
+    vars = vcat(edge_inputs(d, op_id, edge_val), edge_output(d, op_id, edge_val))
+    types = vcat(rule.src_types, rule.res_type)
+    return any(map(zip(vars, types)) do (var, type)
```
Comment (lukem12345) [inline src/acset.jl:549]: Is the zip necessary here?

Comment (GeorgeR227) [reply src/acset.jl]: No but it seems to only work without the zip if I remove the parantheses on the vars after the `do`.

Comment (lukem12345) [reply src/acset.jl]: Ok, this can get changed on the next commit

Comment (GeorgeR227) [reply src/acset.jl]: I actually already changed it in https://github.com/AlgebraicJulia/DiagrammaticEquations.jl/pull/84/commits/b4ba78701f36e31791f862dffc3e0f5d216fa296.

File: src/acset.jl [hunk 4]
```diff
@@ -466,36 +512,105 @@ function infer_sum_types!(d::SummationDecapode, Σ_idx::Int)
   return applied
 end
 
-function apply_inference_rule_op1!(d::SummationDecapode, op1_id, rule)
-  score_src = (rule.src_type == d[d[op1_id, :src], :type])
-  score_tgt = (rule.tgt_type == d[d[op1_id, :tgt], :type])
+"""
+    check_operator(d::SummationDecapode, op_id, rule, edge_val; check_name::Bool = false, check_aliases::Bool = false, ignore_infers::Bool = false, ignore_usertypes::Bool = false)
```
Comment (jpfairbanks) [inline src/acset.jl:516]: docstring formatting

File: test/language.jl [hunk 5]
```diff
@@ -856,13 +842,37 @@ end
...
+    C::Form0
+
+    B == fake(A)
+    D == fake2(A, C)
   end
+  op_not_in_rules_res = deepcopy(op_not_in_rules)
+  infer_types!(op_not_in_rules)
+
+  @test op_not_in_rules == op_not_in_rules_res
+  @test op_not_in_rules !== op_not_in_rules_res
```
Comment (jpfairbanks) [inline test/language.jl:859]: how can both of these tests pass?

File: test/language.jl [hunk 6]
```diff
@@ -1034,83 +1155,12 @@ end
...
-      summand = [3, 17]
-      summation = [1, 1]
-      sum = [5]
-      op1 = [:∂ₜ, :d₀, :d₀, :♯ᵖᵖ, :mag, :⋆₁, :dual_d₁, :⋆₀⁻¹]
-      op2 = [:*, :-, :^, :∧₁₀, :^, :∧₁₀]
-      type = [:Form0, :Form1, :Constant, :Form0, :infer, :Form1, :Form1, :Form1, :Form1, :Form0, :Form0, :PVF, :Form1, :infer, :Literal, :Form0, :Literal, :DualForm1, :DualForm2]
-      name = [:h, :Γ, :n, :ḣ, :sum_1, Symbol("•2"), Symbol("•3"), Symbol("•4"), Symbol("•5"), Symbol("•6"), Symbol("•7"), Symbol("•8"), Symbol("•9"), Symbol("•10"), Symbol("1"), Symbol("•11"), Symbol("2"), Symbol("•_6_1"), Symbol("•_6_2")]
-    end
-  end
-
```
Comment (jpfairbanks) [inline test/language.jl:1113]: Do we not need to test this anymore?

Comment (GeorgeR227) [reply test/language.jl:1113]: It was figured out that this Decapode was actually poorly typed and thus not a good test for ensuring the overloading/inferring was working correctly. Instead I've converted this into a type-checking test. It's been shifted around I but not deleted.

File: test/language.jl (hunk 5 above)
Comment (GeorgeR227) [reply test/language.jl:859]: Line 858 is checking that the ACSets have the same entries, Line 859 is checking that the memory locations are not the same.

File: src/acset.jl [hunk 7]
```diff
@@ -440,6 +442,50 @@ function filterfor_ec_types(types::AbstractVector{Symbol})
...
+
+function bin_broad_arith_ops(op_name)
+  all_ops = map(t -> Operator{Symbol}(t, [t, t], op_name), FORM_TYPES)
+  for type in vcat(USER_TYPES, NUMBER_TYPES)
+    append!(all_ops, map(t -> Operator{Symbol}(t, [t, type], op_name), FORM_TYPES))
+    append!(all_ops, map(t -> Operator{Symbol}(t, [type, t], op_name), FORM_TYPES))
+  end
+
+  all_ops
+end
```
Comment (lukem12345) [inline src/acset.jl:484]: This looks like it is trying to emulate the style of having an abstract type for an "Operator", can this idiom be used explicitly.

**Comment by lukem12345** [comment]
@quffaro This PR checks types via a notion of a “score” which [counts](https://github.com/AlgebraicJulia/DiagrammaticEquations.jl/pull/84/files#diff-193c1fdaed9c36d8a8ce0b7e35147ca00bddc1b93ae40c511596a82a4675730cR527) the number of mis-matched types between a function’s type signature vs. a collection of type signatures (“rules”). It seems to be converging on [some variation](https://github.com/AlgebraicJulia/DiagrammaticEquations.jl/pull/84/files#diff-193c1fdaed9c36d8a8ce0b7e35147ca00bddc1b93ae40c511596a82a4675730cR605) of the “pattern matching” idiom. (If a pattern is fully matched, or no partial match is found (i.e. fall through), throw an error. Else when a partial match is found, throw an error.)

And so I am concerned that we may be unwittingly getting into maintaining a bespoke type checker by way of reimplementing of a pattern matching system. I think that the explicit notion of “scoring” is a step in the right direction, but ultimately should be realized by making use of a fully-featured pattern matching capabilities offered by MLStyle or a similar package like Moshi.

Can I get your thoughts on this? @quffaro

File: src/acset.jl [hunk 8]
```diff
@@ -440,6 +439,88 @@ function filterfor_ec_types(types::AbstractVector{Symbol})
...
+        types1 = vcat(rule1.res_type, rule1.src_types)
+        types2 = vcat(rule2.res_type, rule2.src_types)
+
+        if length(types1) != length(types2)
+          continue
+        end
+
+        score = mapreduce(+, types1, types2; init = 0) do type1, type2
+          if type1 == type2
+            return 0
```
Comment (lukem12345) [inline src/acset.jl:506]: This function should be unpacked. It can return an Int, a Float, or a Bool

File: src/acset.jl [hunk 9]
```diff
@@ -440,6 +439,88 @@ function filterfor_ec_types(types::AbstractVector{Symbol})
...
+    append!(all_ops, map(t -> Operator{Symbol}(t, [t, type], op_name), FORM_TYPES))
+    append!(all_ops, map(t -> Operator{Symbol}(t, [type, t], op_name), FORM_TYPES))
+  end
+
+  all_ops
+end
+
+# TODO: This could probably be implemented using a better version of `check_operator`
+# TODO: Add printing of rules which are ambigious with each other
+function check_rule_ambiguity(type_rules::AbstractVector{Operator{Symbol}})
```
Comment (lukem12345) [inline src/acset.jl:488]: This function is a little difficult to parse. Can this be broken down into component functions?

**Comment by GeorgeR227** [comment]
I've rebased this PR and tests seem to be running. However, I've had to remove this rule, 
```Operator(:DVF, :Form1, :♯ᵖᵈ, [:♯])```
as including it caused some ambiguity tests to fail, since this rule is also included,
```Operator(:PVF, :Form1, :♯ᵖᵖ, [:♯])```.

The main reason is the alias along with the input type. For example, if in the Decapode, I'm given an input with `:Form1` and the operator is ```♯```, I'd have no idea if I should choose the output to be `:DVF` or `:PVF`. I think there are similar problems with the wedge that we get around by not having these generic aliases.

Luckily, this failing shows that the functionality of the rule ambiguity checking is working exactly as intended!

**Comment by lukem12345** [comment]
Yes the notion of "ambiguity" as implemented in code here is a little too strict, since there may be multiple valid discretizations for an operator.

**Comment by GeorgeR227** [comment]
Yeah, I think you've hit on the solution which really lies in overhauling the type inference/operator overloading.

Like you said, we could make the order we apply rules in more concrete, like by assigning some priority. Thus ambiguity could also check similar rules don't have the same priorities. Then an issue with this is that we'd have to make sure there's overall consistency. For example, if in a Decapode we have `:Form1` -> `♯` -> `infer` -> `♭ᵈᵖ` -> `infer`, then we might make a mistake if we are inferring through the `♯` and we assigned higher priority to `:♯ᵖᵖ` instead of `:♯ᵖᵈ`. 

I think in this case we could run only rules that are completely unambiguous and if a rule is ambiguous, we still include those in the running but don't apply it, with the hope that there might be enough information later to disambiguate it. So essentially we no longer run a greedy rule application but check all rules that may apply and play it safe by only applying if there is a single result. So in the above, if we later run the rule for `♭ᵈᵖ` and infer all the types above, then there is no more ambiguity. Then if, when we do type inference/operator overloading, the only rules left to apply are ambiguous or don't apply at all, we can stop there.

I think also valid for the flat and sharp is to make sharp_flat operator, `♭♯`, a supported operator and in this case choose the underlying implementation that works the best.

**Comment by lukem12345** [comment]
Yes a two-staged process, each greedy, sounds best. The first stage of rules that must be applied, and the second of those ambiguous rules which agree in degree but not primality. This two step process can be repeated until no further rules are applied. Luckily the existing infrastructure is already centered around applying sets of rules, so this is purely a higher-level scheduling abstraction.
